File: boj/Stack/BOJ_17298.py
# 백준 17298 오큰수
# 21.05.03

import sys
import heapq
from functools import cmp_to_key
from collections import deque
from itertools import combinations
from itertools import permutations
# sys.setrecursionlimit(10**6)
rd = sys.stdin.readline
INF = int(1e9)

n = int(rd())

A = list(map(int, rd().split()))
stack = []

for i in range(n):
    now = A[i]

    while stack and A[stack[-1]] < now:
        A[stack.pop()] = now
    stack.append(i)

for i in stack:
    A[i] = -1

answer = ""
for i in A:
    answer += str(i) + " "

print(answer)

File: boj/Stack/BOJ_17298.swift
// 백준 17298 오큰수
// 21.05.03

import Foundation

let n = Int(readLine()!)!
var A = readLine()!.split(separator: " ").map { Int(String($0))! }
var stack = [Int]() // 수열 A의 index 담기

for i in stride(from: 0, to: n, by: 1) {
  // 스택의 탑이 A[i]보다 작으면 A[i]는 NGE
  while !stack.isEmpty && A[stack.last!] < A[i]{
    A[stack.removeLast()] = A[i]
  }

  stack.append(i)
}

for i in stack {
  A[i] = -1
}

var answer = ""
for i in A{
  answer += "\(i) "
}

print(answer)